addons/stock/models/product.py

```diff
@@ -307,7 +307,7 @@ def _get_domain_locations_new(self, location_ids):
         # this optimizes [('location_id', 'child_of', locations.ids)]
         # by avoiding the ORM to search for children locations and injecting a
         # lot of location ids into the main query
-        paths_domain = expression.OR([[('parent_path', '=like', loc.parent_path + '%')] for loc in locations])
+        paths_domain = expression.OR([[('parent_path', '=like', loc.parent_path + '%')] for loc in locations]) if locations else []
         loc_domain = [('location_id', 'any', paths_domain)]
         dest_loc_domain = [('location_dest_id', 'any', paths_domain)]
```

Comment:
I think is clearer and probably more correct to just do nothing if there is no `locations`
```suggestion
        if locations:
            paths_domain = expression.OR([[('parent_path', '=like', loc.parent_path + '%')] for loc in locations])
            loc_domain = [('location_id', 'any', paths_domain)]
            dest_loc_domain = [('location_dest_id', 'any', paths_domain)]
```